components/identity-mgt/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/listener/TenantSyncListener.java

```diff
@@ -0,0 +1,401 @@
+        String username = Utils.replaceSystemProperty(
+                identityEventListenerConfig.getProperties().getProperty(TenantManagement.USER_NAME));
+
+        String notificationEndpoint = identityEventListenerConfig.getProperties()
+                .getProperty(TenantManagement.NOTIFICATION_ENDPOINT);
+
+        // Any headers will be defined in the config as header. prefex
+        HashMap<String, String> headers = new HashMap<String, String>();
+        for (Map.Entry<Object, Object> propertiesEntry : identityEventListenerConfig.getProperties().entrySet()) {
+            String key = (String) propertiesEntry.getKey();
+            String value = (String) propertiesEntry.getValue();
+            if (key.startsWith(TenantManagement.HEADER_PROPERTY)) {
+                headers.put(key.split(TenantManagement.HEADER_PROPERTY)[1], value);
+            }
+        }
+        EventRunner eventRunner = new EventRunner(notificationEndpoint, username, password, headers, eventDTO);
+        EXECUTOR.execute(eventRunner);
+
+    }
+
+    protected TenantManagementEventDTO buildPayload(TenantInfoBean tenantInfo, String type, String eventURI,
+                                                    String serverURL) {
+
+        TenantManagementEventDTO eventDTO = new TenantManagementEventDTO();
+
+        eventDTO.setIss(serverURL);
+        eventDTO.setJti(UUID.randomUUID().toString());
+        eventDTO.setIat(System.currentTimeMillis() / 1000L);
+
+        Map<String, TenantManagementEventDTO.EventDetail> events = new HashMap<>();
+
+        TenantManagementEventDTO.EventDetail createEventDetail = new TenantManagementEventDTO.EventDetail();
+        createEventDetail.setInitiatorType(TenantManagement.EVENT_INITIATOR);
+        createEventDetail.setAction(type);
+
+        TenantManagementEventDTO.Tenant tenant = new TenantManagementEventDTO.Tenant();
+        tenant.setId(Integer.toString(tenantInfo.getTenantId()));
+        tenant.setDomain(tenantInfo.getTenantDomain());
+        tenant.setRef(serverURL + "/api/server/v1/tenants/" + tenantInfo.getTenantId());
+
+        if (TenantManagement.EVENT_CREATE_TENANT_URI.equals(eventURI)
+                || TenantManagement.EVENT_UPDATE_TENANT_URI.equals(eventURI)) {
+            List<TenantManagementEventDTO.Owner> owners = new ArrayList<>();
+            TenantManagementEventDTO.Owner owner = new TenantManagementEventDTO.Owner();
+
+            if (TenantManagement.EVENT_CREATE_TENANT_URI.equals(eventURI)) {
+                // set username only it is a creation request
+                owner.setUsername(tenantInfo.getAdmin());
+            }
+            owner.setPassword(tenantInfo.getAdminPassword());
+            owner.setEmail(tenantInfo.getEmail());
+            owner.setFirstname(tenantInfo.getFirstname());
+            owner.setLastname(tenantInfo.getLastname());
+            owners.add(owner);
+            tenant.setOwners(owners);
```

Comment:
We recommend to write the model classes in builder pattern [1], so others cannot update the model object in any other places. Otherwise it'll more error prone.

[1] https://github.com/wso2/carbon-identity-framework/blob/v7.8.392/components/webhook-mgt/org.wso2.carbon.identity.webhook.management/src/main/java/org/wso2/carbon/identity/webhook/management/api/model/Webhook.java